### PR TITLE
merge the ResolveJwksURI into refresh() function

### LINKED ISSUE #30261: failed to fetch jwks_uri on first attempt does not recover
**Bug description**
We setup istio with requestauthentication resource to validate jwt tokens. The issuer is a URL which causes istiod to try to the OIDC discovery of the well known endpoint to retrieve the JWKS. If an error occurs in the first attempt during the istiod initialization phase it won't recover ever. This is because the failed attempt is stored also for upcoming istiod attempts to resolve the JWKS.

**Logs**
```
2020-12-31T14:44:52.737506Z	warn	model	Failed to get jwks_uri for issuer "https://$domain/oauth2/default": Get "https://$domain/oauth2/default/.well-known/openid-configuration": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
2020-12-31T14:44:52.737482Z	error	model	Failed to fetch jwks_uri from "https://$domain/oauth2/default/.well-known/openid-configuration": Get "https://$domain/oauth2/default/.well-known/openid-configuration": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

2020-12-31 16:02:32	0.0.0.0_8080: Issuer 'https://$domain/oauth2/default' in jwt_authn config has invalid local jwks: Jwks RSA [n] or [e] field is missing or has a parse error
2020-12-31T15:02:31.992986Z	error	Failed to fetch jwt public key from "": Get "": unsupported protocol scheme ""
2020-12-31T15:02:31.992972Z	error	model	Failed to fetch public key from "": Get "": unsupported protocol scheme ""
2020-12-31T15:02:31.991866Z	error	Failed to fetch jwt public key from "": Get "": unsupported protocol scheme ""
2020-12-31T15:02:31.991851Z	error	model	Failed to fetch public key from "": Get "": unsupported protocol scheme ""
```

In the istiod code here is where the problem starts:
https://github.com/istio/istio/blob/080374a27dd16646be2649ab377af49183667ac1/pilot/pkg/model/jwks_resolver.go#L261

```go	
       body, err := r.getRemoteContentWithRetry(issuer+openIDDiscoveryCfgURLSuffix, networkFetchRetryCountOnMainFlow)
	if err != nil {
		log.Errorf("Failed to fetch jwks_uri from %q: %v", issuer+openIDDiscoveryCfgURLSuffix, err)
		return "", err
	}
```

Here I think it should either retry or bail out (panic) because it will store the empty string on error and will result in the errors you can see above.
https://github.com/istio/istio/blob/080374a27dd16646be2649ab377af49183667ac1/pilot/pkg/model/jwks_resolver.go#L214

[ ] Docs
[x] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[x] User Experience
[ ] Developer Infrastructure
[ ] Upgrade

**Steps to reproduce the bug**
Reproduce it with having a first attempt failure as described above.

**Version** (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version --short` if you used Helm)

```
$ kubectl version --short
Client Version: v1.18.6
Server Version: v1.18.9-eks-d1db3c
$ istioctl version
client version: 1.8.1
control plane version: 1.8.1
data plane version: 1.8.1 (6 proxies)
```

**How was Istio installed?**
istioctl

**Environment where the bug was observed (cloud vendor, OS, etc)**
AWS eks but it's not environment specific.

### PR DESCRIPTION
**Pull request by domcyrus**
As suggested by @yangminzhu this PR merges the ResolveJwksURI to the refresh() function. Fixes #30261 

[x] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[x] Policies and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request.

[x] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Comment by domcyrus** [comment]
@yangminzhu I think the PR does now looks better and all the unit tests pass, linting is fine and also the issue with a race condition is fixed. There seems to be the following issues still present:
- release-notes_istio (which I tried to fix with /release-note-none cmd)
- two integ-security checks fail

I'm a bit lost on the integration tests but will try to make them run locally. If anything obvious is wrong and someone could give me some pointers I would appreciate it.

**Comment by domcyrus** [comment]
@yangminzhu sorry to have bothered you. I've figured out the issues... I'll need to do some further refactoring. I'll update ASAP.

**Comment by domcyrus** [comment]
@yangminzhu I think now the PR is ready for review. The only thing I was not able to do is to add the release-notes-none label.

File: pilot/pkg/model/jwks_resolver.go [hunk 1]
```diff
@@ -106,11 +99,14 @@ type jwtPubKeyEntry struct {
 	lastUsedTime time.Time
 }
 
+// jwtKey is a key in the JwksResolver keyEntries map.
```
Comment (yangminzhu) [inline pilot/pkg/model/jwks_resolver.go:102]: could you also update the comment of `keyEntries` in the `JwksResolver` struct.

File: pilot/pkg/model/jwks_resolver.go [hunk 2]
```diff
@@ -142,10 +144,11 @@ func init() {
 }
 
 // NewJwksResolver creates new instance of JwksResolver.
-func NewJwksResolver(evictionDuration, refreshInterval, retryInterval time.Duration) *JwksResolver {
+func NewJwksResolver(evictionDuration, refreshDefaultInterval, refreshIntervalOnFailure, retryInterval time.Duration) *JwksResolver {
```
Comment (yangminzhu) [inline pilot/pkg/model/jwks_resolver.go:147]: As you're already here, could you rename this to newJwksResolver()? I think it's not used anywhere out of jwks_resolver.go and jwks_resolver_test.go. Thanks.

File: releasenotes/notes/30261.yaml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: networking
+issue:
+- 30261
+releaseNotes:
+- |
+  **Fixed**  Periodically resolve jwks_uri.
```
Comment (yangminzhu) [inline releasenotes/notes/30261.yaml:8]: could you reword this as a new feature and also include a link to the github issue?

File: releasenotes/notes/30261.yaml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: networking
```
Comment (yangminzhu) [inline releasenotes/notes/30261.yaml:3]: the area should be security, and the kind should be feature

File: pilot/pkg/model/jwks_resolver_test.go [hunk 5]
```diff
@@ -293,14 +202,20 @@ func TestGetPublicKeyUsingTLSBadCert(t *testing.T) {
...
 	}
 }
 
 func TestGetPublicKeyUsingTLSWithoutCABundles(t *testing.T) {
-	r := newJwksResolverWithCABundlePaths(JwtPubKeyEvictionDuration, JwtPubKeyRefreshInterval, testRetryInterval, []string{})
+	r := newJwksResolverWithCABundlePaths(
+		JwtPubKeyEvictionDuration,
+		JwtPubKeyRefreshInterval,
+		testRetryInterval,
+		JwtPubKeyRefreshIntervalOnFailure,
```
Comment (yangminzhu) [inline pilot/pkg/model/jwks_resolver_test.go:216]: Could you add a test case to verify the refresh interval will be switched back to the normal value (20min) from the failure case (1min)?

File: pilot/pkg/model/jwks_resolver.go [hunk 6]
```diff
@@ -340,21 +339,27 @@ func (r *JwksResolver) refresher() {
 	for {
 		select {
 		case <-r.refreshTicker.C:
-			r.refresh()
+			refreshInterval := r.refresh()
+			// update refresh interval on change.
+			if r.refreshInterval != refreshInterval {
+				r.refreshTicker.Stop()
+				r.refreshTicker = time.NewTicker(refreshInterval)
```
Comment (yangminzhu) [inline pilot/pkg/model/jwks_resolver.go:346]: This means if the jwks server is not working for some reason, we will keep sending the request every 1 minutes until it works, will this cause too much traffic to the server or puts too much burden on istiod? Not sure if we need to increase the interval after some failed attempt. @incfly  @howardjohn wdyt?

**Review by yangminzhu — COMMENTED** [review]
Sorry for the late reply, the PR looks good overall, just a few small comments about test case and the retry interval on failure. thanks!

File: releasenotes/notes/30261.yaml (hunk 3 above)
Comment (domcyrus) [reply releasenotes/notes/30261.yaml:8]: Ah I thought the:
```
issue:
- 30261
```
Is already providing the information on the issue without providing a link in the release notes.

File: pilot/pkg/model/jwks_resolver.go (hunk 6 above)
Comment (domcyrus) [reply pilot/pkg/model/jwks_resolver.go:346]: Otherwise we could also use an exponential backoff retry mechanism. Starting with 1 minute and then increase it until it reaches the default interval in case it keeps failing.

**Comment by domcyrus** [comment]
@yangminzhu I'm really sorry but somehow I messed up the updating of the branch and it seemed to have pulled in a lot of other commits which shouldn't belong into this. I would like to close this PR and start with a new one. I'll sync all our conversation into the new PR. Is this ok with you?
